trieste/acquisition/sampler.py

```diff
@@ -276,3 +279,205 @@ def sample(self, at: TensorType, *, jitter: float = DEFAULTS.JITTER) -> TensorTy
+            theta_posterior_mean, theta_posterior_chol_covariance
+        )
+
+    def _prepare_theta_posterior_in_gram_space(self) -> tfp.distributions.Distribution:
+        r"""
+        Calculate the posterior of theta (the feature weights) in the gram space.
+
+         .. math:: \theta \sim N(\Phi^TG^{-1}y,I_m - \Phi^TG^{-1}\Phi)
+
+        where the [n,n] gram matrix :math:`G=(\Phi\Phi^T + \sigma^2I_n)` is defined for the [n,m]
+        matrix of feature evaluations across the training data :math:`\Phi` and
+        observation noise variance :math:`\sigma^2`.
+
+        :return: The posterior distribution for theta.
+        """
+
+        phi = self._feature_functions(self._dataset.query_points)  # [n, m]
+        G = tf.matmul(phi, phi, transpose_b=True)  # [n, n]
+        s = self._noise_variance * tf.eye(self._num_data, dtype=phi.dtype)
+        L = tf.linalg.cholesky(G + s)
+        L_inv_phi = tf.linalg.triangular_solve(L, phi)  # [n, m]
+        L_inv_y = tf.linalg.triangular_solve(L, self._dataset.observations)  # [n, 1]
+
+        theta_posterior_mean = tf.tensordot(tf.transpose(L_inv_phi), L_inv_y, [[-1], [-2]])[
+            :, 0
+        ]  # [m,]
+        theta_posterior_covariance = tf.eye(self._num_features, dtype=phi.dtype) - tf.tensordot(
+            tf.transpose(L_inv_phi), L_inv_phi, [[-1], [-2]]
+        )  # [m, m]
+        theta_posterior_chol_covariance = tf.linalg.cholesky(theta_posterior_covariance)  # [m, m]
+
+        return tfp.distributions.MultivariateNormalTriL(
+            theta_posterior_mean, theta_posterior_chol_covariance
+        )
+
+    def get_trajectory(self) -> Callable[[TensorType], TensorType]:
+        """
+        Generate an approximate function draw (trajectory) by sampling weights
+        and evaluating the feature functions.
+
+        If this is the first call, then we calculate the posterior distributions for
+        the feautre weights.
+
+        :return: A function representing an approximate trajectory from the Gaussian process,
+            taking an input of shape `[N, D]` and returning shape `[N, 1]`
+        """
+
+        if not self._pre_calc:
+            self._feature_functions = RandomFourierFeatures(
+                self._kernel, self._num_features, dtype=self._dataset.query_points.dtype
+            )
+
+            if (
+                self._num_features < self._num_data
+            ):  # if m < n  then calculate posterior in design space (an m*m matrix inversion)
+                self._theta_posterior = self._prepare_theta_posterior_in_design_space()
+            else:  # if n < m  then calculate posterior in gram space (an n*n matrix inversion)
+                self._theta_posterior = self._prepare_theta_posterior_in_gram_space()
+
+            self._pre_calc = True
+
+        def trajectory(x: TensorType) -> TensorType:
+            feature_evaluations = self._feature_functions(x)  # [N, m]
+            theta_sample = self._theta_posterior.sample(1)  # [1, m]
```

Comment:
A stupid question: is it make sense to move `self._theta_posterior` outside `trajectory`? In that sense, the `trajectory` can be a deterministic trajectory sample, which can be optimized for instance.

Reply:
You are right! That was always my intension.

Reply:
Good spot

Reply:
I will fix this and add a test to check that the trajectories are deterministic